rethinkdb/datadog_checks/rethinkdb/metrics/config.py

```diff
@@ -0,0 +1,58 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+from typing import Iterator
+
+from ..connections import Connection
+from ..queries import QueryEngine
+from ..types import Metric
+
+logger = logging.getLogger(__name__)
+
+
+def collect_config_totals(engine, conn):
+    # type: (QueryEngine, Connection) -> Iterator[Metric]
+    """
+    Collect aggregated metrics about cluster configuration.
+
+    See: https://rethinkdb.com/docs/system-tables/#configuration-tables
+    """
+    logger.debug('collect_config_totals')
+
+    totals = engine.query_config_totals(conn)
+    logger.debug('config_totals totals=%r', totals)
+
+    yield {
```

Comment:
The "everything is an iterator of metrics" is a nice interface to develop against, IMO. Also `yield` provides lazy evaluation (improved memory usage).

> `yield` makes debugging harder

How so? (Genuine question! When developing the tracebacks always pinpointed to the exact place where I made a mistake, so I'm not sure what you're referring to…)

Reply:
> How so? (Genuine question! When developing the tracebacks always pinpointed to the exact place where I made a mistake, so I'm not sure what you're referring to…)

When you use a debugger, if it's a list, you can see the content straight away, but if it's an iterator, it's a bit harder.

You can read the value, but have a side effect of consuming the iterator. Usually, while debugging you don't want that.

---

List having a greater API surface and lazy evaluation are good points. It's probably just a matter of style with each way it's own tradeoffs.

Reply:
@AlexandreYang Oh OK, I understand the debugger argument. FWIW, there are debug logs on each yield, so when running in DEBUG mode the output is very thorough already, eg:

```console
$ ddev env start --dev rethinkdb py38
$ ddev env check -l debug rethinkdb py38
...
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/check.go:72 in runCheck) | Running python check rethinkdb rethinkdb:5942258e6ff90a33
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:90) | check config=Config(host='localhost', port=28015, user='datadog-agent', password='*****', tls_ca_cert=None)
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:50) | connected connection_tags=['host:localhost', 'port:28015', 'server:server0', 'proxy:false']
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (config.py:21) | collect_config_totals
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (config.py:24) | config_totals totals={'databases': 1, 'secondary_indexes_per_table': {'heroes': 1}, 'servers': 3, 'tables_per_database': {'doghouse': 1}}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.server.total', 'value': 3, 'tags': []}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.database.total', 'value': 1, 'tags': []}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.database.table.total', 'value': 1, 'tags': ['database:doghouse']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.table.secondary_index.total', 'value': 1, 'tags': ['table:heroes']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:21) | collect_cluster_statistics
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:24) | cluster_statistics stats={'id': ['cluster'], 'query_engine': {'client_connections': 1, 'clients_active': 1, 'queries_per_sec': 13.990213551999998, 'read_docs_per_sec': 0, 'written_docs_per_sec': 0}}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.cluster.queries_per_sec', 'value': 13.990213551999998, 'tags': []}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.cluster.read_docs_per_sec', 'value': 0, 'tags': []}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.cluster.written_docs_per_sec', 'value': 0, 'tags': []}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:57) | collect_server_statistics
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:60) | server_statistics server={'cache_size_mb': 'auto', 'id': '7de33d73-5a19-42dc-851d-eca1aa7991fa', 'name': 'server1', 'tags': ['default', 'primary', 'us']} stats={'id': ['server', '7de33d73-5a19-42dc-851d-eca1aa7991fa'], 'query_engine': {'client_connections': 0, 'clients_active': 0, 'queries_per_sec': 0, 'queries_total': 0, 'read_docs_per_sec': 0, 'read_docs_total': 4, 'written_docs_per_sec': 0, 'written_docs_total': 4}, 'server': 'server1'}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.client_connections', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.clients_active', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.queries_per_sec', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.queries_total', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.read_docs_per_sec', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.read_docs_total', 'value': 4, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.written_docs_per_sec', 'value': 0, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.written_docs_total', 'value': 4, 'tags': ['server:server1', 'default', 'primary', 'us']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:60) | server_statistics server={'cache_size_mb': 'auto', 'id': '82c6e502-94d5-4c1c-97d1-4fb6df5bcbbd', 'name': 'server2', 'tags': ['default', 'eu']} stats={'id': ['server', '82c6e502-94d5-4c1c-97d1-4fb6df5bcbbd'], 'query_engine': {'client_connections': 0, 'clients_active': 0, 'queries_per_sec': 0, 'queries_total': 0, 'read_docs_per_sec': 0, 'read_docs_total': 1, 'written_docs_per_sec': 0, 'written_docs_total': 4}, 'server': 'server2'}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.client_connections', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.clients_active', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.queries_per_sec', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.queries_total', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.read_docs_per_sec', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.read_docs_total', 'value': 1, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.written_docs_per_sec', 'value': 0, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'monotonic_count', 'name': 'rethinkdb.stats.server.written_docs_total', 'value': 4, 'tags': ['server:server2', 'default', 'eu']}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | - | (statistics.py:60) | server_statistics server={'cache_size_mb': 'auto', 'id': 'be596951-e580-49b7-99d4-37ba69b0e9dd', 'name': 'server0', 'tags': ['default', 'us']} stats={'id': ['server', 'be596951-e580-49b7-99d4-37ba69b0e9dd'], 'query_engine': {'client_connections': 1, 'clients_active': 1, 'queries_per_sec': 14.777860751999999, 'queries_total': 102, 'read_docs_per_sec': 0, 'read_docs_total': 0, 'written_docs_per_sec': 0, 'written_docs_total': 0}, 'server': 'server0'}
2020-03-13 18:05:27 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:122 in LogMessage) | rethinkdb:5942258e6ff90a33 | (rethinkdb.py:70) | submit_metric metric={'type': 'gauge', 'name': 'rethinkdb.stats.server.client_connections', 'value': 1, 'tags': ['server:server0', 'default', 'us']}
...
```

Do you think that'd enough in terms of debugging aid? These have been very helpful for me while developing the integration (and they'll also be output in any debug flares provided by customers).

Reply:
> Do you think that'd enough in terms of debugging aid?

I was referring to interactive debugging using `pdb` or debuggers from IDEs.

But good logging like this will help too of course.

Reply:
> I was referring to interactive debugging using pdb or debuggers from IDEs.

I see, yes…

So, the main reason I'm a bit reluctant to switch to a list style is readability.

I don't know WYT, but to me each syntax-highlighted `yield` visually corresponds to "we're submitting a metric here" very clearly, which makes the code pretty easy to read. OTOH with lists we'd need to pattern-match for `metrics.append()` or `metrics.extend()`, making the code harder to read.

```python
yield {...}

# vs

metrics.append({...})
```

At this point we're nitpicking of course, but I really like this streaming style so I'd like to find a compromise…

Would aggregating into a single list in `DefaultBackend.collect_metrics()` be enough? One could easily dig through the `metrics` list (one group of metrics at a time) by putting breakpoints there:

```python
def collect_metrics(self, conn):
    # type: (Connection) -> List[Metric]
    # NOTE: aggregate into a list to facilitate interactive debugging.
    metrics = []  # type: List[Metric]

    for collect in self.collect_funcs:
        for metric in collect(self._query_engine, conn):
            metrics.append(metric)

    return metrics
```

Reply:
We are probably nitpicking. It's just two different ways of implementing the same thing with it's own trade-offs already mentioned above :) So it's fine if we keep things that way.